web_infrastructure/jenkins_plugin.py

```diff
@@ -0,0 +1,781 @@
+                # Make the checksum of the currently installed plugin
+                md5sum_old = hashlib.md5(
+                    open(plugin_file, 'rb').read()).hexdigest()
+
+            if self.params['version'] in [None, 'latest']:
+                # Take latest version
+                plugin_url = (
+                    "%s/latest/%s.hpi" % (
+                        self.params['updates_url'],
+                        self.params['name']))
+            else:
+                # Take specific version
+                plugin_url = (
+                    "{0}/download/plugins/"
+                    "{1}/{2}/{1}.hpi".format(
+                        self.params['updates_url'],
+                        self.params['name'],
+                        self.params['version']))
+
+            if (
+                    self.params['updates_expiration'] == 0 or
+                    self.params['version'] not in [None, 'latest'] or
+                    md5sum_old is None):
+
+                # Download the plugin file directly
+                r = self.__download_plugin(plugin_url)
+
+                # Write downloaded plugin into file if checksums don't match
+                if md5sum_old is None:
+                    # No previously installed plugin
+                    self.__write_file(plugin_file, r)
+                    changed = True
+                else:
+                    # Get data for the MD5
+                    data = r.read()
+
+                    # Make new checksum
+                    md5sum_new = hashlib.md5(data).hexdigest()
+
+                    # If the checksum is different from the currently installed
+                    # plugin, store the new plugin
+                    if md5sum_old != md5sum_new:
+                        self.__write_file(plugin_file, data)
+                        changed = True
+            else:
+                # Check for update from the updates JSON file
+                plugin_data = self.__download_updates()
+
+                try:
+                    sha1_old = hashlib.sha1(open(plugin_file, 'rb').read())
+                except Exception:
+                    e = get_exception()
+                    self.module.fail_json(
+                        msg="Cannot calculate SHA1 of the old plugin.",
+                        details=e.message)
+
+                sha1sum_old = base64.b64encode(sha1_old.digest())
+
+                # If the latest version changed, download it
+                if sha1sum_old != plugin_data['sha1']:
+                    r = self.__download_plugin(plugin_url)
+                    self.__write_file(plugin_file, r)
+                    changed = True
+
+        return changed
+
+    def __download_updates(self):
+        updates_file = 'update-center.json'
+        updates_tmp_file = "%s/%s" % (tempfile.gettempdir(), updates_file)
+        download_updates = True
+
+        # Check if we need to download new updates file
+        if os.path.isfile(updates_tmp_file):
```

Comment:
Can this be made to work without caching the updates file across module runs?

We generally avoid saving state or caching files on the remote host between module runs, as it tends to make the module more prone to unexpected failures or behavior:

- Different remote users running the module may not all have access to the cached file, which would cause module failure. If they do, the permissions are probably too open, allowing unintended access to the file.
- Different playbooks may have different configuration for using the cached file (in this case the expiration).
- If the cached file is corrupted, the module will be unable to recover until the cached file is expired.
- The cache file could be corrupted if there are two concurrent runs of the module on the remote system.

Some of these issues could be solved by:

- Caching the file in the remote user's home directory instead of the system-wide temp dir.
- Writing the file to a temp location and verifying it can be parsed before saving it for later use.

However, it would be better if the caching could be avoided all together.